src/Export/DynamicProductGroupService.php

```diff
@@ -218,10 +246,109 @@ private function buildCriteria(): Criteria
+
+        return $this->cache->getItem($id);
+    }
+
+    private function setTotalInCache(CacheItemInterface $cacheItem, $total): void
+    {
+        $cacheItem->set($total);
+        $cacheItem->expiresAfter(self::CACHE_LIFETIME_PRODUCT_GROUP);
+        $this->cache->save($cacheItem);
+    }
+
+    private function getDynamicGroupsTotalCacheItem(): CacheItemInterface
     {
-        $id = sprintf('%s_%s', self::CACHE_ID_PRODUCT_GROUP, $this->shopkey);
+        $id = sprintf('%s_%s_total', self::CACHE_ID_PRODUCT_GROUP, $this->shopkey);
 
         return $this->cache->getItem($id);
     }
+
+    /**
+     * Gets the total count of dynamic product groups from cache
+     */
+    public function getDynamicProductGroupTotalFromCache(): int
+    {
+        $cacheItem = $this->getDynamicGroupsTotalCacheItem();
+        if ($cacheItem && $cacheItem->isHit()) {
+            $cacheItem->expiresAfter(self::CACHE_LIFETIME_PRODUCT_GROUP);
+            $this->cache->save($cacheItem);
+
+            return (int)$cacheItem->get();
+        }
+
+        return 0;
+    }
+
+    /**
+     * If we have reached the last page of the dynamic product group export, we set a flag in cache to
+     * know that the dynamic product groups are warmed up
+     */
+    public function setDynamicProductGroupWarmedUpFlag(int $total = 0): void
+    {
+        if (($this->start + $this->count) >= $total) {
+            $cacheItem = $this->getDynamicProductGroupWarmedUpCacheItem();
+            if ($cacheItem) {
+                $cacheItem->set(true);
+                $cacheItem->expiresAfter(self::CACHE_LIFETIME_PRODUCT_GROUP);
+                $this->cache->save($cacheItem);
+            }
+        }
+    }
+
+    /**
+     * Gets the total count of available dynamic product groups to be exported
+     */
+    private function getTotalDynamicProductGroupsCount(): int
+    {
+        $total = 0;
+        $criteria = $this->buildCriteria();
+        $categories = $this->getCategoriesWithProductGroups($criteria);
+        foreach ($categories as $categoryEntity) {
+            $productStream = $categoryEntity->getProductStream();
+            if (!$productStream) {
+                continue;
+            }
+
+            $total++;
+        }
+
+        return $total;
+    }
+
+    private function getCategoriesWithProductGroups(Criteria $criteria): ?CategoryCollection
+    {
+        /** @var CategoryCollection $categories */
+        $categories = $this->categoryRepository->search($criteria, $this->context)->getEntities();
+
+        if ($categories === null || !$categories->count()) {
+            return null;
+        }
+
+        return $categories;
+    }
+
+    public function getDynamicProductGroupWarmedUpCacheItem(): CacheItemInterface
+    {
+        $id = sprintf('%s_%s_dpg_warmup', self::CACHE_ID_PRODUCT_GROUP, $this->shopkey);
+
+        return $this->cache->getItem($id);
+    }
+
+    public function isDynamicProductGroupWarmedUp(): bool
```

Comment:
Good idea of moving this as a responsibility of the service, but the reason we didn't do this is because that stuff like the `ProductStreamBuilder` may not be available for older Shopware versions. Have you ensured that this still works properly?

Reply:
Yes because we are passing the warm up flag as true if the version is lower than 6.3 so the warm up logic is not called to trigger the product stream builder.

Reply:
:+1: